---
title: "Applying CCN to an existing taxonomy"
output: html_notebook
---
  
# Overview and preparations
  
This script will allow easy application of the Common Cell type Nomenclature (CCN) to an existing hierarchically-structured taxonomy of cell types.  , An early verion of the CCN is available at the [Allen Brain Map](https://portal.brain-map.org/explore/classes/nomenclature), and the updated version presented here [is available on arXiv](https://arxiv.org/abs/2006.05406) and is undergoing review at eLife.  This script requires a taxonomy and some cell meta-data as input and outputs several files useful for publication and for ingest into in-process taxonomy services.  Please post any thoughts about the CCN to the [Community Forum](https://community.brain-map.org/c/cell-taxonomies/)!  
  
Prior to running the scripts, the follow steps need to be performed:  
  
1. Install R on your computer
2. Install these libraries: `dplyr`, `dendextend`, `data.table`, and `ggplot2` in R.
3. Download `required_scripts.R` to your working directory and put in a **scripts** subfolder.  
4. Create an **data** subfolder and put the dendrogram and metadata files to annotated therein (see next section for details).
5. *(Optional)* Download `dend.RData`, `nomenclature_table.csv`, and `cell_metadata.csv` as an example taxonomy for annotation (see our [Allen Institute Transcriptomics Explorer](https://celltypes.brain-map.org/rnaseq/human/cortex)) and put in a **data** subfolder.
6. *(Optional)* Install RStudio on your computer.
7. *(Optional)* Install the `jsonlite` library if you want to save the final dendrogram in json format.
  
If you choose to install R studio on your computer, `build_annotation_tables.rmd` can be opened and run directly from RStudio.  If you would prefer to run this script from R, the code blocks in this document can be directly copied into R in order and run.  
  
There are two files required as input to this script (that should go into the **data** subfolder):  

1. `dend.RData`: this is an R `dendrogram` object representing the taxonomy to be annotated.  If you used R for cell typing in your manuscript, this is likely a variable that was created at some point during this process and from which your dendrogram images are made.  While this assumes a hierarchical structure of the data, additional cell sets of any kind can be made later in the script.  Code for converting from other formats to the R `dendrogram` format is not provided, but please post to the [Community Forum](https://community.brain-map.org/c/cell-taxonomies/) if you have questions about this.
2. `cell_metadata.csv`: a table which includes a unique identifier for each cell (in this example it is stored in the `sample_name` column) as well as the corresponding cell type from the dendrogram for each cell (in this example it is stored in the `cluster_label` column).  Additional metadata of any kind can be optionally included in this table.  
We provide files for the taxonomy from [(Hodge et al 2019)](https://www.nature.com/articles/s41586-019-1506-7) as an example.  
  
The general steps of this script are as follows.  First, a unique taxonomy_id is chosen, which will be used as a prefix for all the cell set accession IDs. The R dendrogram is read in and used as the starting point for defining cell sets by including both provisional cell types (terminal leaf nodes) and groups of cell types with similar expression patterns (internal nodes). The main script then assigns accession ids and other requred columns for each cell set and outputs an intermediate table, along with a minimally annotated dendrogram for visualization. Next, the user **manually annotates these cell sets** to include common usage terms (aligned aliases), and can also manually add additional cell sets which can correspond to any combination of cell types in the taxonomy. This updated table is read back into R and dendrograms are optionally updated to include the new nomenclature information. Next, cells are assigned nomenclature tags corresponding to their cell set assignments.  This is automated for any cell sets based on cell types or other available meta-data.  Finally, the code produces a set of standardized files for visualization of updated taxonomic structure and for input into in-process databases for cross-taxonomy comparison (described below) or inclusion as supplemental materials for manuscripts utilizing the annotated taxonomy.  
  
  
# Build the nomenclature

The remainder of this script describes how to run the CCN in R.  At this point open RStudio (or R) and start running the code below.  The first few blocks correpsond to housekeeping things needed to get the workspace set up.
  
  
### Workspace setup
  
```{r setup}
# NOTE: REPLACE THIS LINK BELOW WITH YOUR WORKING DIRECTORY
outputFolder = "\\\\allen/programs/celltypes/workgroups/hct/cellTaxonomy/nomenclature/upload_to_CTTS/marmoset_M1/"

setwd(outputFolder)                           # Needed only if copying and pasting in R
knitr::opts_chunk$set(echo = TRUE)            # Needed only for RStudio
knitr::opts_knit$set(root.dir = outputFolder) # Needed only for RStudio
```
  
  
### Load required libraries  
  
```{r}
suppressPackageStartupMessages({
  library(dplyr)
  library(dendextend)
  library(ggplot2)
  library(data.table)
  library(jsonlite)  # optional
})
```
  
  
### Load the accessory scripts  
  
This file includes all of the scripts that are required for the CCN, and is well-annotated.  Remember to put it in the **scripts** subfolder!  
  
```{r}
source("required_scripts.R")  # Additional required files
options(stringsAsFactors = FALSE)
```
  
  
### Read in the updated nomenclature
  
This file is the one just created above.  If needed change the link to the file below.  
  
```{r}
# REPLACE THIS LINE OF CODE WITH CODE TO READ IN YOUR UPDATED NOMENCLATURE TABLE, IF NEEDED
updated_nomenclature <- read.csv("nomenclature_table_CCN201912132.csv")
```
  
  
Create an additional tag called `child_cell_set_accessions`, which is a "|"-separated character vector indicating all of the child set sets in the dendrogram (e.g., "provisional cell types", "leaves", or "terminal nodes").  This is calculated by using the `cell_set_label` tags and will help with integration into downstream ontology use cases.  
  
```{r}
updated_nomenclature <- define_child_accessions(updated_nomenclature)
# We then rewrite the table above, which should be identical, but with a new column for child cell sets
write.csv(updated_nomenclature,"nomenclature_table_CCN201912132_withExtras.csv",row.names=FALSE)
```
  
  
### Update the dendrogram

As discussed above, we have provided a dendrogram of human MTG cell types (called "dend") as an example.  Any dendrogram of cell types in the "dendrogram" R format, where the desired cluster aliases are in the "labels" field of the dendrogram will work for this code.  Other formats *might* work and will try to be forced into dendrogram format.  

```{r, fig.width=12, fig.height=5}
# REPLACE THIS LINE OF CODE WITH CODE TO READ IN YOUR DENDROGRAM, IF NEEDED
dend <- readRDS("dend.RData")

# Attempt to format dendrogram if the input is in a different format
dend <- as.dendrogram(dend)
#dend <- prune(dend,"Oligo L3-6 OPALIN LRP4-AS1")
```
  
  
The entire script for assigning all nomenclature is done as a single function.  If you'd prefer to run it line by line (for example if your data is in slightly different format), see the `build_nomenclature_table` function in the `required_scripts.R` file.  This function has been reasonably-well commented to try and explain how each section works.  
  
```{r}
first_label <- setNames("RNAseq", 1)
nomenclature_information <- build_nomenclature_table(
  dend,  first_label, "taxonomy_id", "taxonomy_author", "taxonomy_citation", "structure", "ontology_tag")
```


This code will take the information from the table above and add it to the initial dendrogram object.  When plotted the only visible difference will be that the new cell set alias names (if any) will show up to replace the n## labels from the initial plot.  However, ALL of the meta-data read in from the table will be added to the relevant nodes or leafs.  Cell sets not linked to the tree will be ignored in this step, but will be added to the relevent text files output below.  
  
```{r}
updated_dendrogram <- update_dendrogram_with_nomenclature(nomenclature_information$initial_dendrogram,updated_nomenclature)

pdf("updated_dendrogram_CCN201912132.pdf",height=8,width=15)
plot_dend(updated_dendrogram, node_size=3)
dev.off()
```
  
### Save the dendrogram in various formats
  
Plots only show a small fraction of the data available in these dendrogram objects; to see the rest the dendrogram needs to be saved.  We find both the R "dendrogram" format and the "json" format useful for different applications at the Allen Institute and code for saving data as both are presented below.  
  
```{r} 
# Save as an R data object
save(updated_dendrogram, file="updated_dendrogram_CCN201912132.RData")
```
  
  
(This section can be skipped if json format is not needed.)  
  
```{r}
# Convert to a list
# NOTE: Only some features of dendrogram can be converted to a list.  If this function 
#       crashes, the "omit_names" variable may need to be updated
dend_list <- dend_to_list(updated_dendrogram, omit_names = c("markers","markers.byCl","class"))

# Save as a json file
dend_JSON <- toJSON(dend_list, complex = "list", pretty = TRUE)
out <- file("dend_CCN201912132.json", open = "w")
writeLines(dend_JSON, out)
close(out)
```
  
  
# Define cell to cell set mappings

Up to this point the document describes how to apply the CCN to cell sets based on a hierarchical (or non-hierarchical) dendrograms, with an additional manual annotation step.  This final section describes how cells within a data set can be mapped onto this nomenclature.  Doing this would better allow mapping of cells and cell sets between multiple taxonomies, particularly in cases where multiple taxonomies contain the same cells.  
  
### Read in meta-data an set up variables
  
Prior to assigning nomenclature to individual cells, we need to read in the meta-data associated with each cell and specify the variables that will allow linking to the updated nomenclature for each cell set.  More specifically, we need to create a character vector of `cell_set_accession_id`s called `cell_id` that corresponds to each cell used for generating the taxonomy.  This variable is used as a starting point to assign all cells to all cell sets.  In this example, we generate this vector by reading in a file of meta-data which includes a `preferred_alias` (called `cluster_label`) corresponding to the cell type names from [(Hodge et al 2019)](https://www.nature.com/articles/s41586-019-1506-7).  
  
```{r} 
# Read in metadata and collect correct columns for sample name and cell set accession id
# REPLACE THIS LINE OF CODE WITH CODE TO READ IN YOUR METADATA, IF NEEDED
metadata  <- read.csv("cell_metadata_CCN201912132.csv")
samples   <- metadata$sample_name

# OPTION 1: COLUMN FOR ACCESSION ID ALREADY EXISTS
# cell_id <- metadata$cell_set_accession

# OPTION 2: NEED TO GENERATE COLUMN FROM DENDROGRAM LABELS
label_col <- "cluster"  # Column name with dendrogram labels
cell_id   <- updated_nomenclature[match(metadata[,label_col],updated_nomenclature$cell_set_preferred_alias),"cell_set_accession"]
cell_id[is.na(cell_id)] = "none"
```
  
  
### Assign dendrogram cell sets
  
Next, we want to automatically link each cell to each cell set that is available in the dendrogram.  This is done as a single line of code (Option 1). *Note: if a dendrogram is not available, the `mapping` table can be set up using a single cell set using Option 2 below.*  
  
```{r} 
# OPTION 1: START FROM A DENDROGRAM
#mapping   <- cell_assignment_from_dendrogram(updated_dendrogram,samples,cell_id)

# OPTION 2: USE ONLY THE `updated_nomenclature` table
mapping <- data.frame(sample_name=samples, call=((cell_id==cell_id[1])+1-1))
colnames(mapping) <- c("sample_name",cell_id[1])
```  
  
The result of this script is a data frame where the first columns corresponds to the cell `sample_name`.  This is the term used at the Allen Institute for a unique cell ID.  This ID is unique across **all data at the Allen Institute**.  In principle, a replacement (or additional) unique ID value could be added for importing into external databases.  The remaining columns corrspond to the probabilities of each cell mapping to each cell type (from the dendrogram).  In this case we define hard probabilities (0 = unassigned to cell set; 1 = assigned to cell set) but this could be adapted to reflect real probabilities calculated elsewhere.  
  
  
### Assign cell type-based cell sets  
  
This section assigns cell sets that were defined as combinations of cell types, but that were NOT included in the above section.  As written, this function requires *assumes that the `cell_set_label` is assigned using the specific format described above*.  If Option 2 was selected in the previous code block, all cell_set_labels must have the same prefix.  
  
```{r} 
mapping <- cell_assignment_from_groups_of_cell_types(updated_nomenclature,cell_id,mapping)
```  
  
Finally, we can add cell to cell set mappings specified by any other metadata.  No function is required for this mapping.  Instead replace the set_cells_to_keep variable as needed to select cells of interest.  In principle, one could also read in text files and add meta-data columns that way, if desired.  **For most taxonomies, this step can be skipped, as in most cases cell sets are defined exclusively as a combination of cell types.**  
  
Output relevant cell set accession ids to the screen for convenience.  
  
```{r} 
print("Remaining cell sets:")
missed_ids    = setdiff(updated_nomenclature$cell_set_accession,colnames(mapping))
print(paste0(missed_ids,collapse="; "))
```
  

Finally, we output the cell to cell_set assignments as a csv file.  
  
```{r} 
fwrite(mapping,"cell_to_cell_set_assignments_CCN201912132.csv")
```    
  
  
Session info.  
  
```{r} 
sessionInfo()
``` 
  